crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -40,24 +46,160 @@ pub use bevy_ecs_macros::States;
 /// }
 ///
 /// ```
-pub trait States: 'static + Send + Sync + Clone + PartialEq + Eq + Hash + Debug + Default {
-    /// The type returned when iterating over all [`variants`](States::variants) of this type.
-    type Iter: Iterator<Item = Self>;
+pub trait States: 'static + Send + Sync + Clone + PartialEq + Eq + Hash + Debug + Default {}
+
+/// Types that can match world-wide states.
+pub trait StateMatcher<S: States>: 'static + Send + Sync + Clone {
+    /// Check whether to match with the current state
+    fn match_state(&self, state: &S) -> bool;
+}
+
+impl<S: States> StateMatcher<S> for S {
+    fn match_state(&self, state: &S) -> bool {
+        self == state
+    }
+}
+
+/// A struct for containing a state matcher function
+#[derive(Clone)]
+pub struct StateMatcherFunction<S: States>(Arc<dyn Fn(&S) -> bool + 'static + Send + Sync>);
```

Comment:
The `Arc` here makes me suspicious: I don't think you should need it.

Reply:
I was getting compile errors when I tried box, but will see if Rc works

Reply:
ah, the issue is that you're trying to make `dyn Fn` clonable, but that's not allowed since Clone isn't object safe. Is it possible to take a `dyn fn` here instead or do we need to be able to capture the environment?

Reply:
Originally I didn't think it would be possible, due to a mistaken idea of when you can create `fn` pointers. Was able to get rid of the whole StateMatcherFunction struct!